UDP Multicast implementation

Sunil-K-S opened the pull request:
Signed-off-by: sunil.s <sunil.s@lge.com>

ssugiura commented on src/daemon/udp_connection/dlt_daemon_udp_socket.c:
```diff
@@ -0,0 +1,262 @@
+
+    if(daemon_local == NULL)
+    {
+        return ret_val;
```
Should return DLT_RETURN_WRONG_PARAMETER

Sunil-K-S replied:
Done

ssugiura commented on src/daemon/udp_connection/dlt_daemon_udp_socket.c:
```diff
@@ -0,0 +1,262 @@
+        g_udp_sock_fd = fd;
+        // set global multicast addr
+        dlt_daemon_udp_setmulticast_addr(daemon_local);
+        ret_val = DLT_RETURN_TRUE;
```
Any reason we need to change ret_val to DLT_RETURN_TRUE here? I think the caller function only checks if it's below 0 or not.

Sunil-K-S replied:
Code modified

ssugiura commented on src/daemon/udp_connection/dlt_daemon_udp_socket.c:
```diff
@@ -0,0 +1,262 @@
+        // set global multicast addr
+        dlt_daemon_udp_setmulticast_addr(daemon_local);
+        ret_val = DLT_RETURN_TRUE;
+        dlt_log(LOG_INFO, " initialize udp socket success\n");
```
Can you change the log level to DEBUG?

Sunil-K-S replied:
Done

ssugiura commented on src/daemon/udp_connection/dlt_daemon_udp_socket.c:
```diff
@@ -0,0 +1,262 @@
+    int enable_reuse_addr = 1;
+    int sockbuffer = DLT_DAEMON_RCVBUFSIZESOCK;
+    char portnumbuffer[SOCKPORT_MAX_LEN] = {0};
+    struct addrinfo hints, *servinfo = NULL, *addrinfo_iterator = NULL;
```
Separate into multiple lines

Sunil-K-S replied:
Done

ssugiura commented on src/daemon/udp_connection/dlt_daemon_udp_socket.c:
```diff
@@ -0,0 +1,262 @@
+
+    if(sock == NULL)
+    {
+        return DLT_RETURN_ERROR;
```
Return DLT_RETURN_WRONG_PARAMETER

Sunil-K-S replied:
Done

ssugiura commented on src/daemon/udp_connection/dlt_daemon_udp_socket.c:
```diff
@@ -0,0 +1,262 @@
+                    clientinfo->clientaddr_size) < 0)
+            dlt_log(LOG_ERR,"dlt_daemon_udp_clientmsg_send2: Send UDP Packet Data failed\n");
+
+    } else {
```
I propose this function to be static so that NULL check would be done in the caller function.

Sunil-K-S replied:
Done

ssugiura commented on src/daemon/udp_connection/dlt_daemon_udp_socket.c:
```diff
@@ -0,0 +1,262 @@
+            free(data);
+            data = NULL;
+        } else {
+            dlt_log(LOG_ERR,"dlt_daemon_udp_clientmsg_send: calloc failure\n");
```
As proposed above, error case should go first and succeeded case afterwards.

Sunil-K-S replied:
Done

ssugiura commented on src/examples/CMakeLists.txt:
```diff
@@ -33,7 +33,14 @@ add_executable(	dlt-example-filetransfer ${dlt_example_filetransfer_SRCS})
 set_target_properties(dlt-example-filetransfer PROPERTIES LINKER_LANGUAGE C)
 
-install(TARGETS dlt-example-user dlt-example-user-func dlt-example-user-common-api dlt-example-filetransfer
+set(dlt-example-multicast-clientmsg-view_SRCS dlt-example-multicast-clientmsg-view.c)
```
This only works if WITH_UDP_CONNECTION is set to ON, right? Then these lines should be put under if (WITH_UDP_CONNECTION) condition.

Sunil-K-S replied:
Done

ssugiura commented on src/examples/dlt-example-multicast-clientmsg-view.c:
```diff
@@ -0,0 +1,195 @@
+     }
+
+
+/**** MODIFICATION TO ORIGINAL */
```
Is this comment needed?

Sunil-K-S replied:
Removed

ssugiura commented on CMakeLists.txt:
```diff
@@ -78,6 +78,7 @@ option(WITH_DLT_LOGSTORAGE_CTRL_UDEV "PROTOTYPE! Set to ON to build logstorage c
 option(WITH_DLT_USE_IPv6      "Set to ON for IPv6 support"                                                       ON)
 option(WITH_DLT_KPI           "Set to ON to build src/kpi binaries"                                              OFF)
 option(WITH_DLT_FATAL_LOG_TRAP "Set to ON to enable DLT_LOG_FATAL trap(trigger segv inside dlt-user library)"    OFF)
+option(WITH_UDP_CONNECTION     "Set to ON to enable dlt UDP multicast SUPPORT"      ON)
```
Add a status message in bottom of this file.

Sunil-K-S replied:
Added

votrungchi commented on src/daemon/udp_connection/dlt_daemon_udp_socket.c:
```diff
@@ -0,0 +1,262 @@
+            continue;
+        }
+
+        if(setsockopt(*sock, SOL_SOCKET, SO_RCVBUF, &sockbuffer, DLT_DAEMON_RCVBUFSIZESOCK)
```
The last parameter is sizeof(sockbuffer) or sizeof(int), not DLT_DAEMON_RCVBUFSIZESOCK.

Sunil-K-S replied:
Done

votrungchi commented on src/daemon/udp_connection/dlt_daemon_udp_socket.c:
```diff
@@ -0,0 +1,262 @@
+                       == SYSTEM_CALL_ERROR)
+        {
+            dlt_vlog(LOG_WARNING, "[%s/%d] Setsockopt error %s\n", __func__, __LINE__,
+                    strerror(errno));
```
close socket here.

Sunil-K-S replied:
Done

votrungchi commented on src/daemon/udp_connection/dlt_daemon_udp_socket.c:
```diff
@@ -0,0 +1,262 @@
+                      == SYSTEM_CALL_ERROR)
+        {
+            dlt_vlog(LOG_WARNING, "[%s/%d] Setsockopt error %s\n", __func__, __LINE__,
+                    strerror(errno));
```
close socket here.

Sunil-K-S replied:
Done

Sunil-K-S commented on src/daemon/CMakeLists.txt:
```diff
@@ -14,6 +14,11 @@
 
 
+if(WITH_UDP_CONNECTION)
+    include_directories(${CMAKE_SOURCE_DIR}/src/daemon/udp_connection)
```
Changed to PROJECT_SOURCE_DIR

Sunil-K-S commented on src/daemon/CMakeLists.txt:
```diff
@@ -14,6 +14,11 @@
 #######
 
 
+if(WITH_UDP_CONNECTION)
```
Done

Sunil-K-S commented on src/daemon/CMakeLists.txt:
```diff
@@ -55,6 +58,11 @@ else()
 endif()
 
+if(WITH_UDP_CONNECTION)
+    set(dlt_daemon_SRCS ${dlt_daemon_SRCS} ${CMAKE_SOURCE_DIR}/src/daemon/udp_connection/dlt_daemon_udp_socket.c)
```
Done

Sunil-K-S commented on src/daemon/dlt-daemon.c:
```diff
@@ -597,6 +604,32 @@ int option_file_parser(DltDaemonLocal *daemon_local)
+                        else {
+                            daemon_local->UDPConnectionSetup = MULTICAST_CONNECTION_DISABLED;
+                            fprintf(stderr,
+                                    "Invalid value for UDPConnectionSetup setto default %ld\n",
```
Done

Sunil-K-S commented on src/daemon/dlt.conf:
```diff
@@ -183,3 +183,16 @@ ControlSocketPath = /tmp/dlt-ctrl.sock
+
+# UDP multicast port(default:3491)
+UDPMulticastIPPort = 3491
+
```
Done

Sunil-K-S commented on src/daemon/dlt.conf:
```diff
@@ -183,3 +183,16 @@ ControlSocketPath = /tmp/dlt-ctrl.sock
 # OfflineLogstorageCacheSize = 30000
+
+##############################################################################
+# UDP Multicast Configuration                                                #
```
Done

Sunil-K-S commented on src/daemon/udp_connection/dlt_daemon_udp_socket.c:
```diff
@@ -0,0 +1,262 @@
+void dlt_daemon_udp_setmulticast_addr(DltDaemonLocal* daemon_local)
+{
+    dlt_daemon_udp_init_clientstruct(&g_udpmulticast_addr);
+//#ifdef DLT_USE_IPv6 // IPv6 Support
```
Comments removed

Sunil-K-S commented on src/daemon/udp_connection/dlt_daemon_udp_socket.c:
```diff
@@ -0,0 +1,262 @@
+// **************************************************************************
+void dlt_daemon_udp_init_clientstruct(DltDaemonClientSockInfo *clientinfo_struct)
+{
+    if(clientinfo_struct != NULL){
```
Code modified

Sunil-K-S commented on src/daemon/udp_connection/dlt_daemon_udp_socket.c:
```diff
@@ -0,0 +1,262 @@
+// Out Param  : NIL
+// Description: set the multicast addr to global variables
+// **************************************************************************
+void dlt_daemon_udp_setmulticast_addr(DltDaemonLocal* daemon_local)
```
Added

Sunil-K-S commented on src/daemon/udp_connection/dlt_daemon_udp_socket.c:
```diff
@@ -0,0 +1,262 @@
+// **************************************************************************
+DltReturnValue dlt_daemon_udp_connection_setup(DltDaemonLocal* daemon_local)
+{
+    int fd = -1;
```
Done

Sunil-K-S commented on src/daemon/dlt_daemon_common.h:
```diff
@@ -104,6 +104,15 @@ extern "C" {
+#ifdef UDP_CONNECTION_SUPPORT
+    #define MULTICASTIPADDRESS "225.0.0.37"
+    #define MULTICASTIPPORT 3491
+    #define MUlticastIP_MAX_SIZE 256
```
Done

Sunil-K-S commented on src/daemon/udp_connection/dlt_daemon_udp_socket.c:
```diff
@@ -0,0 +1,262 @@
+    PRINT_FUNCTION_VERBOSE(verbose);
+    if((clientinfo != NULL) && (clientinfo->isvalidflag == ADDRESS_VALID) &&
+        (data1 != NULL) && (size1 > 0) &&
+        (data2 != NULL) && (size2 > 0)) {
```
Code modified currently.
Added with the intention that when UDP Buffer is implemented data2 would be expected to be NULL as the data comes from buffer directly. In that case data2 should not be checked or NULL

votrungchi replied:
Sorry, but I still can not get your idea? First of first, data2 still be checked NULL in code. 
And my original comment is that we should support only send data1 (data2 can be omit/NULL).

Sunil-K-S replied:
As per comments by @ssugiura, the function is made static and hence the checks for NULL  have been removed(checked in caller function).
Regarding the data1 & data2 : 
        data1 consists of the header information of the dlt-message 
        data2 constsits of databuffer (payload) information of the dlt-message
Hence, the log is sent only if both is available.

Sunil-K-S commented on src/daemon/udp_connection/dlt_daemon_udp_socket.c:
```diff
@@ -0,0 +1,262 @@
+
+    snprintf(portnumbuffer, SOCKPORT_MAX_LEN, "%d", servPort);
+    if ((getaddrinfo_errorcode = getaddrinfo(NULL, portnumbuffer, &hints, &servinfo)) != 0) {
+        dlt_vlog(LOG_WARNING, "[%s/%d] getaddrinfo: %s\n", __func__, __LINE__,
```
Done

Sunil-K-S commented:
> Sorry that I added too much comments. Other than that, I have two more requests:
> 
> 1. Please run uncrustify to harmonize the coding style as suggested in README.
> 2. Add summary of WITH_UDP_CONNECTION in README. Under General options should be fine.
> 
> The feature itself looks good for me.



> Sorry that I added too much comments. Other than that, I have two more requests:
> 
> 1. Please run uncrustify to harmonize the coding style as suggested in README.
> 2. Add summary of WITH_UDP_CONNECTION in README. Under General options should be fine.
> 
> The feature itself looks good for me.

Summary has now been added.
Uncrustify has been run on the modified code and the latest code is pushed.
